File: cotizacion.py
from enum import Enum as PyEnum
from sqlalchemy.sql import func
from src.models.user import db # Assuming db is initialized in a central models.py or user.py
# Ensure other related models are imported if direct relationships are defined here
# from src.models.cliente import Cliente
# from src.models.user import User

class EstadoCotizacionEnum(PyEnum):
    CREADA = "Creada"
    ENVIADA = "Enviada"
    ACEPTADA = "Aceptada"
    EN_EJECUCION = "En Ejecucion"
    ENTREGADA = "Entregada"
    RECHAZADA = "Rechazada"
    MODIFICADA = "Modificada" # For new versions or updates
    CANCELADA = "Cancelada"

class Cotizacion(db.Model):
    __tablename__ = "cotizaciones"

    id_cotizacion = db.Column(db.Integer, primary_key=True, autoincrement=True)
    numero_cotizacion = db.Column(db.String(50), nullable=False, unique=True)
    id_cliente = db.Column(db.Integer, db.ForeignKey("clientes.id_cliente"), nullable=False)
    id_usuario_creador = db.Column(db.Integer, db.ForeignKey("usuarios.id_usuario"), nullable=False)
    
    fecha_emision = db.Column(db.Date, nullable=False, default=func.current_date())
    fecha_validez = db.Column(db.Date)
    titulo_asunto = db.Column(db.String(255))
    
    estado = db.Column(db.Enum(EstadoCotizacionEnum), nullable=False, default=EstadoCotizacionEnum.CREADA)
    
    subtotal_neto = db.Column(db.Numeric(15,2), default=0.00)
    descuento_general_porcentaje = db.Column(db.Numeric(5,2), default=0.00)
    monto_descuento_general = db.Column(db.Numeric(15,2), default=0.00)
    monto_neto_final = db.Column(db.Numeric(15,2), default=0.00) # subtotal_neto - monto_descuento_general
    monto_iva = db.Column(db.Numeric(15,2), default=0.00)
    monto_total_cotizacion = db.Column(db.Numeric(15,2), default=0.00) # monto_neto_final + monto_iva
    
    forma_pago_descripcion = db.Column(db.Text)
    tiempo_entrega_estimado = db.Column(db.String(255))
    validez_oferta_descripcion = db.Column(db.Text)
    notas_adicionales = db.Column(db.Text)
    
    fecha_creacion_registro = db.Column(db.TIMESTAMP, server_default=func.now())
    fecha_ultima_modificacion = db.Column(db.TIMESTAMP, server_default=func.now(), onupdate=func.now())

    # Relationships
    cliente = db.relationship("Cliente", backref=db.backref("cotizaciones", lazy=True))
    usuario_creador = db.relationship("User", backref=db.backref("cotizaciones_creadas", lazy=True))
    items = db.relationship("ItemCotizacion", backref="cotizacion", lazy="dynamic", cascade="all, delete-orphan")
    pagos = db.relationship("PagoCliente", backref="cotizacion", lazy="dynamic", cascade="all, delete-orphan")
    # proyecto = db.relationship("Proyecto", backref="cotizacion_origen", uselist=False, cascade="all, delete-orphan") # If one-to-one

    def __repr__(self):
        return f'<Cotizacion {self.numero_cotizacion}>'

File: cotizacion.dart
import 'package:hive/hive.dart';
import 'package:uuid/uuid.dart';
import 'cliente.dart';
import 'item.dart';

part 'cotizacion.g.dart';

@HiveType(typeId: 3)
class Cotizacion extends HiveObject {
  @HiveField(0)
  final String id;

  @HiveField(1)
  String numeroCotizacion;

  @HiveField(2)
  String idCliente; // Referencia al ID del cliente

  @HiveField(3)
  DateTime fechaEmision;

  @HiveField(4)
  DateTime? fechaValidez;

  @HiveField(5)
  String? tituloAsunto;

  @HiveField(6)
  String estado; // "Creada", "Enviada", "Aceptada", "En Ejecucion", "Entregada"

  @HiveField(7)
  List<Item> items;

  @HiveField(8)
  double subtotalNeto;

  @HiveField(9)
  double descuentoGeneralPorcentaje;

  @HiveField(10)
  double montoDescuentoGeneral;

  @HiveField(11)
  double montoNetoFinal;

  @HiveField(12)
  double montoIva;

  @HiveField(13)
  double montoTotalCotizacion;

  @HiveField(14)
  String? formaPagoDescripcion;

  @HiveField(15)
  String? tiempoEntregaEstimado;

  @HiveField(16)
  String? validezOfertaDescripcion;

  @HiveField(17)
  String? notasAdicionales;

  @HiveField(18)
  DateTime fechaCreacionRegistro;

  @HiveField(19)
  DateTime fechaUltimaModificacion;

  Cotizacion({
    String? id,
    required this.numeroCotizacion,
    required this.idCliente,
    required this.fechaEmision,
    this.fechaValidez,
    this.tituloAsunto,
    this.estado = 'Creada',
    required this.items,
    this.subtotalNeto = 0.0,
    this.descuentoGeneralPorcentaje = 0.0,
    this.montoDescuentoGeneral = 0.0,
    this.montoNetoFinal = 0.0,
    this.montoIva = 0.0,
    this.montoTotalCotizacion = 0.0,
    this.formaPagoDescripcion,
    this.tiempoEntregaEstimado,
    this.validezOfertaDescripcion,
    this.notasAdicionales,
    DateTime? fechaCreacionRegistro,
    DateTime? fechaUltimaModificacion,
  }) : 
    this.id = id ?? const Uuid().v4(),
    this.fechaCreacionRegistro = fechaCreacionRegistro ?? DateTime.now(),
    this.fechaUltimaModificacion = fechaUltimaModificacion ?? DateTime.now();

  // Método para calcular todos los totales
  void calcularTotales(double ivaPorcentaje) {
    // Calcular subtotal sumando el valor total de cada ítem
    subtotalNeto = 0.0;
    for (var item in items) {
      subtotalNeto += item.valorTotalItem;
    }

    // Calcular descuento general
    montoDescuentoGeneral = subtotalNeto * (descuentoGeneralPorcentaje / 100);

    // Calcular monto neto final
    montoNetoFinal = subtotalNeto - montoDescuentoGeneral;

    // Calcular IVA
    montoIva = montoNetoFinal * (ivaPorcentaje / 100);

    // Calcular total
    montoTotalCotizacion = montoNetoFinal + montoIva;

    // Actualizar fecha de modificación
    fechaUltimaModificacion = DateTime.now();
  }

  // Método para crear una copia con cambios
  Cotizacion copyWith({
    String? numeroCotizacion,
    String? idCliente,
    DateTime? fechaEmision,
    DateTime? fechaValidez,
    String? tituloAsunto,
    String? estado,
    List<Item>? items,
    double? subtotalNeto,
    double? descuentoGeneralPorcentaje,
    double? montoDescuentoGeneral,
    double? montoNetoFinal,
    double? montoIva,
    double? montoTotalCotizacion,
    String? formaPagoDescripcion,
    String? tiempoEntregaEstimado,
    String? validezOfertaDescripcion,
    String? notasAdicionales,
  }) {
    return Cotizacion(
      id: this.id,
      numeroCotizacion: numeroCotizacion ?? this.numeroCotizacion,
      idCliente: idCliente ?? this.idCliente,
      fechaEmision: fechaEmision ?? this.fechaEmision,
      fechaValidez: fechaValidez ?? this.fechaValidez,
      tituloAsunto: tituloAsunto ?? this.tituloAsunto,
      estado: estado ?? this.estado,
      items: items ?? List.from(this.items),
      subtotalNeto: subtotalNeto ?? this.subtotalNeto,
      descuentoGeneralPorcentaje: descuentoGeneralPorcentaje ?? this.descuentoGeneralPorcentaje,
      montoDescuentoGeneral: montoDescuentoGeneral ?? this.montoDescuentoGeneral,
      montoNetoFinal: montoNetoFinal ?? this.montoNetoFinal,
      montoIva: montoIva ?? this.montoIva,
      montoTotalCotizacion: montoTotalCotizacion ?? this.montoTotalCotizacion,
      formaPagoDescripcion: formaPagoDescripcion ?? this.formaPagoDescripcion,
      tiempoEntregaEstimado: tiempoEntregaEstimado ?? this.tiempoEntregaEstimado,
      validezOfertaDescripcion: validezOfertaDescripcion ?? this.validezOfertaDescripcion,
      notasAdicionales: notasAdicionales ?? this.notasAdicionales,
      fechaCreacionRegistro: this.fechaCreacionRegistro,
      fechaUltimaModificacion: DateTime.now(),
    );
  }

  // Método para convertir a Map (útil para exportar/importar)
  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'numeroCotizacion': numeroCotizacion,
      'idCliente': idCliente,
      'fechaEmision': fechaEmision.toIso8601String(),
      'fechaValidez': fechaValidez?.toIso8601String(),
      'tituloAsunto': tituloAsunto,
      'estado': estado,
      'items': items.map((item) => item.toMap()).toList(),
      'subtotalNeto': subtotalNeto,
      'descuentoGeneralPorcentaje': descuentoGeneralPorcentaje,
      'montoDescuentoGeneral': montoDescuentoGeneral,
      'montoNetoFinal': montoNetoFinal,
      'montoIva': montoIva,
      'montoTotalCotizacion': montoTotalCotizacion,
      'formaPagoDescripcion': formaPagoDescripcion,
      'tiempoEntregaEstimado': tiempoEntregaEstimado,
      'validezOfertaDescripcion': validezOfertaDescripcion,
      'notasAdicionales': notasAdicionales,
      'fechaCreacionRegistro': fechaCreacionRegistro.toIso8601String(),
      'fechaUltimaModificacion': fechaUltimaModificacion.toIso8601String(),
    };
  }

  // Método para crear desde Map
  factory Cotizacion.fromMap(Map<String, dynamic> map) {
    return Cotizacion(
      id: map['id'],
      numeroCotizacion: map['numeroCotizacion'],
      idCliente: map['idCliente'],
      fechaEmision: DateTime.parse(map['fechaEmision']),
      fechaValidez: map['fechaValidez'] != null ? DateTime.parse(map['fechaValidez']) : null,
      tituloAsunto: map['tituloAsunto'],
      estado: map['estado'],
      items: (map['items'] as List).map((item) => Item.fromMap(item)).toList(),
      subtotalNeto: map['subtotalNeto'],
      descuentoGeneralPorcentaje: map['descuentoGeneralPorcentaje'],
      montoDescuentoGeneral: map['montoDescuentoGeneral'],
      montoNetoFinal: map['montoNetoFinal'],
      montoIva: map['montoIva'],
      montoTotalCotizacion: map['montoTotalCotizacion'],
      formaPagoDescripcion: map['formaPagoDescripcion'],
      tiempoEntregaEstimado: map['tiempoEntregaEstimado'],
      validezOfertaDescripcion: map['validezOfertaDescripcion'],
      notasAdicionales: map['notasAdicionales'],
      fechaCreacionRegistro: DateTime.parse(map['fechaCreacionRegistro']),
      fechaUltimaModificacion: DateTime.parse(map['fechaUltimaModificacion']),
    );
  }

  @override
  String toString() {
    return 'Cotizacion(id: $id, numeroCotizacion: $numeroCotizacion, estado: $estado, montoTotal: $montoTotalCotizacion)';
  }
}
